### DISCUSSION TITLE
x-skip directive

### DISCUSSION THREAD

**Discussion by TibboddiT** [alpinejs/alpine#567] [4. General]
I'm currently building a complex app, where each page consists in a gigantic alpine component with hundreds of directives (yes, i really like alpinejs).

But at any given time, only a small subset of the reactive elements is visible. You can picture it like a hierarchy of nested tabs, with a lot of nested x-for. To improve performance, I decided to implement a directive which, when encountered during the component update tree walk, skip entirely the subtree of the affected element.

I named it `x-skip`, and I use it like this:
```html
<template x-for="(tab, index) in tabs" :key="index">
  <div x-bind:class="{ 'active': tab.isActive }" x-skip="!tab.isActive">
    <div>
      <!-- complex content with lot of directives we can skip if this tab is not visible -->
    </div>
  </div>
</template>
```

Since this addition is, I believe, very minimal, and resulted in a dramatic performance increase for my use case, I was wondering if I should submit a PR...

What do you think ? Is such a directive something consistent with the framework's minimalism ?

**Comment by HugoDF** [comment]
This is an interesting use-case

We had a PR for an `x-ignore` directive that works like your example I think that was closed.

If we it's a wide enough use case I think we can make a case for it.

**Comment by SimoTod** [comment]
It's worth noting that you can achieve the same result using x-if

**Comment by TibboddiT** [comment]
Using x-if was indeed my first Idea, but in this use case i generally dont want elements to be removed from the DOM. For example i could have initialized some complex js library on elements that need to stay there to not having to initialize it again when their visibility will change . I use x-skip when I know the (enormous) subtree of an element doesn't need to be updated. I don't want to waste resources walking thousands of elements, containing hundreds with x-attrs, when i'm sure no changes in the component state will cause a change deeper in the DOM.

I will provide an example of such use case to illustrate the impact of the complete walk of ’updateElements’ on a component with a big DOM tree.

About the x-ignore directive proposal, I had the feeling it was more about security, and iirc it wasn't something that could be conditional: if an element has an x-ignore attr, it's inner HTML will be ignored forever. For my use case, deep elements *might* need an update to reflect a change, even if it's not very often.

**Reply by SimoTod** [reply]
Fair points

**Reply by Saeven** [reply]
If you do open a PR, you've got my vote!

**Comment by Saeven** [comment]
Was this idea ever developed?  I really need something like this as well.  Here's an oversimplified use case (bad one, but enough to demonstrate):

```
<div x-data="{ prices: { 1: { price: '$2' }, 2: { price: '$5'}}, chosen: null}">
  <button x-on:click="chosen = 1">Buy Simple</button>  
  <button x-on:click="chosen = 2">Buy Advanced</button>  
  <hr>
  <!-- would like to ignore this next line if nothing is chosen -->
  <p x-show="chosen > 0">Your Price is <span x-text="prices[chosen].price"></span></p>  
</div>
```

If I could conditionally ignore a block full of evaluations, it would be a big performance enhancement for me.  This example above 'works', but not without dumping a ton of errors since prices[null] does not exist when first loaded.

**Reply by danddanddand** [reply]
@Saeven This is exactly how x-if works. Simply next your `<p>` tag in a `<template x-if=“chosen”> </template>`

**Comment by MuzafferDede** [comment]
I was working on a pr to improve x-ignore with expression such as `x-ignore="expression"` but there is some deep conditions and directive order to figure out. @calebporzio might have better perspective on this. I was hoping that `:x-ignore="expression" would work too but unfortunately it doesn't. For now you can use as @danddanddand  mentioned.

```html
<p x-show="chosen > 0">Your Price is 
  <template x-if="chosen">
    <span x-text="prices[chosen].price"></span>
  </template>
</p>  
```

**Reply by danddanddand** [reply]
@MuzafferDede I feel like I'm missing something, isn't x-if doing exact same thing that x-ignore + expression would do? I mean I suppose if you really didn't want the stuff evaluated you could just `<template x-if="conditions"><div x-ignore><!-- nested--></div></template>` but 99.9% of things wouldn't be very expensive to add to the dom and interoperate with alpine would they? 

We have some pretty complicated form modals fetched from server and rendered once success finishes and it's negligible time to render vs get/post -> server side time -> response (which is still not detectable to user). We also did some performance testing on V3 recently and time between alpine:init and alpine:initialized was rarely over 400ms and that was with 1000's of components (basically our largest, most complicated pages). We even have some table listings where in desktop view you hover elements it'll show pretty large preview fetched from server and moving your mouse around the listing quickly (triggering a bunch of them to appear/vanish) is extremely fast although each are fetched from server when mouse enter occurs. 

I should also note we're hitting 98% performance scores on almost all pages in our application and we have not tuned performance much. 

> ![Screen Shot 2021-06-25 at 10 54 07 PM](https://user-images.githubusercontent.com/47586871/123500067-483dee00-d609-11eb-8c9c-7ebd369280dc.png)

That particular page has 224 components:

> ![Screen Shot 2021-06-25 at 11 07 05 PM](https://user-images.githubusercontent.com/47586871/123500199-4a547c80-d60a-11eb-84a8-5618afa76937.png)

Anyway, can you elaborate on use case for this @MuzafferDede, just trying to learn/understand.

**Reply by KevinBatdorf** [reply]
@TibboddiT brought up on Discord that you may need the nodes in the DOM for other parts of your application or for any number of reasons.

Here's 3 ways to do it now. Alpine will still walk the dom though so something like `x-ignore` with a condition would be nice for that.

https://codepen.io/KevinBatdorf/pen/NWpQZgV?editors=1010

**Reply by danddanddand** [reply]
@KevinBatdorf Oh, ya that makes sense ... I didn't think of that

**Reply by danddanddand** [reply]
@KevinBatdorf Btw, I love your examples. Keep up the great work!

**Comment by MuzafferDede** [comment]
@danddanddand there is an order of running directives : 

```js
let directiveOrder = [
    'ignore',
    'ref',
    'data',
    'bind',
    'init',
    'for',
    'model',
    'transition',
    'show',
    'if',
    DEFAULT,
    'element',
]
```
which means if you use `x-if` on an element and if there is other directives until `x-if` they will evaluated. That's why `x-ignore` is on top to save this cost. But currently i am not sure if there is a way to evaluate `x-ignore`'s condition to apply so we can save the rest of the evaluation.

**Reply by danddanddand** [reply]
I get it now. I'm all for it now that I understand use case.

**Comment by TibboddiT** [comment]
Hello, this idea was indeed developed :) But i didn't took the time to make a proper PR for it, sorry.
The implementation wasn't so complex, but is now fairly outdated, and cannot easily be ported to 3.x.

The good news is that adding this directive to 3.x is even easier :)

Here is how it works :

- when an x-skip directive is found by the walker on an element of the tree, if evaluated to a truthy value, a property `_x_skip` is added and set to `true` on all children of this element, else `_x_skip` is set to `false`
- the walker always test elements to handle to check if  `_x_skip` property exists on it and is true, and if this is the case, stop walking, the same way it does for `x-ignore`

If you all are ok with this, I will send a PR.

I have the feeling that `x-ignore` could be modified slightly to even eliminate the need for an `x-skip` directive, but I'm not sure as I'm not yet fluent with 3.x.

**Reply by MuzafferDede** [reply]
@TibboddiT i had the same feeling and i took one step forward and work on it. I think it has to be done at `x-ignore` since it is the first directive that is evaluated. To me, described idea of `x-skip` is re-inventing the `x-ignore`. 

My current expectation is somehow get `x-ignore` with expression evaluated so `x-data` evaluated before `x-ignore` or a way to evaluate `x-ignore`'expression before all other directives so this could work as:

```html
<div x-data="{ 
        show: false, 
        arr: [], 
        run() {
            this.arr.push({foo: 'Hello World'})
        } 
    }">
    <div x-ignore="!show" x-text="arr[0].foo"></div>
    <button @click="run()">Run</button>
</div>
```
@KevinBatdorf any idea about getting `x-data` evaluated properties while `x-ignore` doing its thing?

**Reply by KevinBatdorf** [reply]
@MuzafferDede Not sure if you get notified of PRs but I put an attempt at solving this here: https://github.com/alpinejs/alpine/pull/1648

It's not perfect as you can't have a custom directive on the same line as `x-ignore`, but I think it's reasonable enough for an edge case like this.

**Reply by MuzafferDede** [reply]
Yes, I checked it. But I think it can't be reverted. I was expecting to see toggling. As you said, this is really an case.

**Reply by KevinBatdorf** [reply]
Oh so you mean something to basically pause rendering? I was under the impression the idea was to defer processing only. 

Once alpine initializes an element, it's a much lower cost after that. It only walks the dom the one time. (Should rephrase that to "it doesn't walk the dom every time theres an update")

Maybe something could just pause the mutation observer. I can check tomorrow

Or maybe a 'pause' helper and directive combo where something like x-pause will keep state and a $unlessPaused() helper to render on condition (identical to the one I posted above)

**Reply by MuzafferDede** [reply]
Maybe this sample will give you my perspective on this:

```html
<div x-data="{
             activeText: 'I am active',
               items: [
                 {active: true},
                 {active: false}, 
                 {active: true}
               ]
             }">
  <template x-for="(item,index) in items" :key="index">
    <span x-ignore="!item.active" x-text="activeText">Nah, i am not active</span>
  </template>
</div>
```
If items are fetch from an api, and there is a condition to check, how would we implement this?

**Reply by KevinBatdorf** [reply]
Maybe it's missing more context but you could just do that with a ternary operator easily enough. That's not an expensive operation either. If there was something complex it could also be memoized.

I think initial load is where you may need to squeeze out performance gains.

**Reply by MuzafferDede** [reply]
Yeah, that's why I said maybe we don't need at all because there is always some workarounds. I remember @bep had some use cases for `x-ignore`, he could give a better samples to figure out if something needs to be done.

**Comment by calebporzio** [comment]
I haven't read through this entire thread, but just some quick thoughts:
x-ignore runs before x-data, so we can't evaluate expressions with it.

I would be pretty hesitant to add an x-skip - between x-if and x-ignore, most cases should be covered - if there are others maybe an x-skip could be a third party package.

I probably need to be reminded the utility of x-sip?

**Reply by KevinBatdorf** [reply]
There's also an issue of custom directives not having priority over `x-ignore` as well. I'm assuming all custom directives are added with the `DEFAULT` priority.

https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L144-L157

I haven't really felt the need for increasing that priority outside of this case though. I suppose at a minimum they would always need to come after `x-data` regardless.

**Reply by MuzafferDede** [reply]
After working on in for a while;

I was able to evaluate the `x-ignore="expression"` in `x-ignore` directive by 

```js
Alpine.nextTick(()=>{
  evaluate(closestRoot(el), expression) // returns evaluated `x-ignore`'s expression 
})
```
and doing this in `effect` was letting me to watch changes but too many new things to figure out with v3. So moving with baby steps. 

But working on this matter feels not right to me and feels like conflicts with `x-ignore`'s purpose. Maybe we shouldn't expect such a thing from `x-ignore` since it is its nature behavior.  

Simple question to ask on this "Can we have a toggleable `x-ignore` ?"

- Yes 
  - Is it expensive?
- No
  - I think we can live with that since there is a lot of workarounds and it is edge case.

**Reply by TibboddiT** [reply]
After digging a little into v3, I really think there is no need for `x-skip` anymore.

The performance problems the directive was solving in v2 were all connected to the `updateElements` function that was called for every mutation of data. This function walk the DOM tree each time and apply the directive handlers, even for gigantic subtrees that were hidden by an `x-show` (think about very complex tabs, `x-if` wasn't an option for hidden ones since I needed elements to stay in the DOM).

But in v3 (unless I miss something) the DOM is not walked anymore each time to update elements. The `effect` technique make the optimization useless.

**Reply by KevinBatdorf** [reply]
It's usefulness would only be on initial render of you have A LOT of elements to walk and set up that you could defer. I think.

But pausing reactivity may be useful for other reasons. I haven't looked into how x-for runs in v3 but that part has always had performance issues under certain conditions, for example.
